teps/0141-platform-context-variables.md

```diff
@@ -221,190 +232,319 @@ The downside of this approach is that Tasks can't easily reference the values of
+to reference platform-provided context variables
+- The mechanism platform builders use to specify the values of each context variable per Run
 
-<!--
-(optional)
+## Design Details
 
-Go in to as much detail as necessary here.
-- What are the caveats to the proposal?
-- What are some important details that didn't come across above?
-- What are the core concepts and how do they relate?
--->
+### Stability level
 
+This feature will start as an alpha feature and will be enabled by a new feature flag "enable-platform-context-variables".
 
-## Design Details
+### TaskRun and PipelineRun syntax
 
-<!--
-This section should contain enough information that the specifics of your
-change are understandable. This may include API specs (though not always
-required) or even code snippets. If there's any ambiguity about HOW your
-proposal will be implemented, this is the place to discuss them.
-
-If it's helpful to include workflow diagrams or any other related images,
-add them under "/teps/images/". It's upto the TEP author to choose the name
-of the file, but general guidance is to include at least TEP number in the
-file name, for example, "/teps/images/NNNN-workflow.jpg".
--->
+This TEP proposes adding support for the context variable substitutions in TaskRuns and PipelineRuns of the form `$(context.platform.foo)`, where `foo` is the name of a parameter supplied by the platform. Platforms may provide string, array, or object parameters, so TaskRuns and PipelineRuns may also use substitutions like `$(context.platform.my-array-param[*])` or `$(context.platform.my-object-param.key)`.
+
+For example, for the clone + build Pipeline referenced above, a user could define the following PipelineRun:
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: PipelineRun
+metadata:
+  name: clone-kaniko-build-push-run
+spec:
+  context:
+    params:
+    - name: repo_url
+      value: https://github.com/tektoncd/pipeline
+    - name: image_registry
+      value: gcr.io/user
+  pipelineRef:
+    name: clone-kaniko-build-push
+  params:
+  - name: repo-url
+    value: $(context.platform.repo_url)
+  - name: image
+    value: $(context.platform.image_registry)/myapp
+  workspaces:
+  - name: source-code
+    ...
+```
+
+These variables will be supported for Tasks defined inline in TaskRuns or Pipelines defined inline in PipelineRuns and can be used in any Task/Pipeline fields that [currently accept variable substitutions](https://github.com/tektoncd/pipeline/blob/main/docs/variables.md#fields-that-accept-variable-substitutions). These variable substitutions will be supported in referenced remote Tasks and Pipelines as well.
+
+For example, the following example is how to reference the platform variables in remote Pipeline:
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Pipeline
+metadata:
+  name: clone-kaniko-build-push
+spec:
+  tasks:
+  - name: fetch-source
+    params:
+    - name: repo-url
+      value: $(context.platform.repo_url)
```

Comment:
where is `repo-url` context defined?

Reply:
That's a great question. It is not defined in `Pipeline` since it is a platform variable, so the pipelines used in this platform can reference to the `repo_url` directly. The feature is not designed for reusable tasks, pipelines, and I have mentioned this in the design evaluation section